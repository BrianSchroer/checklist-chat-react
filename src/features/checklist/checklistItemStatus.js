export const NOT_STARTED                = 'NotStarted';
export const IN_PROGRESS                = 'InProgress';
export const IN_PROGRESS_WITH_ISSUES    = 'InProgressWithIssues';
export const COMPLETED_SUCCESSFULLY     = 'CompletedSuccessfully';
export const COMPLETED_WITH_ISSUES      = 'CompletedWithIssues';
export const COMPLETED_WITH_ERRORS      = 'CompletedWithErrors';
export const CANCELED                   = 'Canceled';
