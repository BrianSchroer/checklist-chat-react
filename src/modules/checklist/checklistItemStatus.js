export const NOT_STARTED                = 'NotStarted';
export const IN_PROGRESS                = 'InProgress';
export const IN_PROGRESS_WITH_ISSUES    = 'InProgressWithIssues';
export const COMPLETED_SUCCESSFULLY     = 'CompletedSuccessfully';
export const COMPLETED_WITH_ISSUES      = 'CompletedWithIssues';
export const COMPLETED_WITH_ERRORS      = 'CompletedWithErrors';
export const CANCELED                   = 'Canceled';

export const options = [
    { value: NOT_STARTED, text: 'Not Started' },
    { value: IN_PROGRESS, text: 'In Progress' },
    { value: IN_PROGRESS_WITH_ISSUES, text: 'In Progress With Issues' },
    { value: COMPLETED_SUCCESSFULLY, text: 'Completed Successfully' },
    { value: COMPLETED_WITH_ISSUES, text: 'Completed with Issues' },
    { value: COMPLETED_WITH_ERRORS, text: 'Completed with Errors' },
    { value: CANCELED, text: 'Canceled' }
];
