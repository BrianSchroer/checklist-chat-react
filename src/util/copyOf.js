export default original => Object.assign({}, original);
