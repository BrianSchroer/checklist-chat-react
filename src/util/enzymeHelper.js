const enzymeHelper = class {

    static nestedFind(wrapper, selector) {
        const selectors = selector.split('>').map(node => node.trim());
        let parent = wrapper;
        let found = [];

        for (let i = 0; i < selectors.length; i++) {
            found = parent.find(selectors[i]);
            if (!found.length) {
                break;
            }
        }

        return found;
    }

    static findSingle(wrapper, selector) {
        const findArray = enzymeHelper.nestedFind(wrapper, selector);

        if (findArray.length != 1) {
            throw(`expected 1 find for selector "${selector}", but found ${findArray.length}.`);
        }

        return findArray.first();
    }
};

export default enzymeHelper;
