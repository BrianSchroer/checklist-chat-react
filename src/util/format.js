import moment from 'moment';

export default class {
    static time(input = '') {

        const isoDateFormatRegex = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}([+-]\d{2}:\d{2})?/g;
        let formatted = input;

        if (input && input.match(isoDateFormatRegex)) {
            formatted = this.dateTimeString(input, 'h:mm a');
        }

        return formatted;
    }

    static dateTimeString(input, formatString) {
        let formatted = input;

        if (input) {
            const testMoment = moment(input);

            if (testMoment.isValid()) {
                const testFormat = testMoment.format(formatString);
                if (testFormat != 'Invalid date') {
                    formatted = testFormat;
                }
            }
        }

        return formatted;
    }
}
