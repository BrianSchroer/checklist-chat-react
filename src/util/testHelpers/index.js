import {shallow, mount} from 'enzyme';
import enzymeHelper from './enzymeHelper';

export {
    shallow as shallow,
    mount as mount,
    enzymeHelper as enzymeHelper
};
