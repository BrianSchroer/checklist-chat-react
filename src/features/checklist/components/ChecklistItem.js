import React, {PropTypes} from 'react';
import ChecklistItemStatusIcon from './CheckllistItemStatusIcon';
import PlannedVsActual from '../../../components/PlannedVsActual';
import format from '../../../util/format';
import ChecklistItemComments from './CheckListItemComments';

// This component returns a tbody because it might require multiple tr's and React requires
// returning a single root element. (It's legal for tables to have multiple tbody's.)

const ChecklistItem = ({checklistItem}) => {
    return (
        <tbody>
            <tr>
                <td><input type="button" value={checklistItem.sequenceNumber}/></td>
                <td><ChecklistItemStatusIcon status={checklistItem.status}/></td>
                <td className="timestamp">
                    <PlannedVsActual
                        planned={format.time(checklistItem.scheduledStartTime)}
                        actual={format.time(checklistItem.actualStartTime)} />
                </td>
                <td className="timestamp">
                    <PlannedVsActual
                        planned={format.time(checklistItem.scheduledEndTime)}
                        actual={format.time(checklistItem.actualEndTime)} />
                </td>
                <td>{checklistItem.description}</td>
                <td>{checklistItem.userName}</td>
                <td>&nbsp;</td>
            </tr>
            {
                checklistItem.chatMessages && checklistItem.chatMessages.length > 0 &&
                    <ChecklistItemComments chatMessages={checklistItem.chatMessages} />
            }
        </tbody>
     );
};

ChecklistItem.propTypes = {
    checklistItem: PropTypes.object.isRequired
};

export default ChecklistItem;
