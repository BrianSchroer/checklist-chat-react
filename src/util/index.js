import format from './format';
import uiHelper from './uiHelper';
import validationHelper from './validationHelper';

export {
    format as format,
    uiHelper as uiHelper,
    validationHelper as validationHelper
};
