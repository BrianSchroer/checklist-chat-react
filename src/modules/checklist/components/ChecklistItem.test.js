import expect from 'expect';
import React from 'react';
import {shallow, enzymeHelper} from '../../../util/testHelpers';
import {format} from '../../../util';
import * as checklistItemStatus from '../checklistItemStatus';
import ChecklistItem from './ChecklistItem';

const defaultChecklistItem = {
    sequenceNumber: 1,
    status: checklistItemStatus.NOT_STARTED,
    description: "item description",
    scheduledStartTime: '2016-12-08T14:00:00.222Z',
    scheduledEndTime: '2016-12-08T15:00:00.222Z',
    actualStartTime: '2016-12-08T14:03:00.222Z',
    actualEndTime: '2016-12-08T15:30:00.222Z',
    userName: 'Jabba Script'
};

function render(checklistItemOverrides = {}) {
    const props = {
        checklistItem: Object.assign({}, defaultChecklistItem, checklistItemOverrides),
        onEditRequest: () => {}
    };

    return shallow(<ChecklistItem {...props} />);
}

describe('ChecklistItem', () => {

    describe('edit button column', () => {
        const selector = 'tbody > tr > td > button.checklist-item-button';

        it('should be rendered with sequence number', () => {
            const button = enzymeHelper.findSingle(render(), selector);
            expect(button.text()).toEqual(defaultChecklistItem.sequenceNumber);
        });

        it('should call onEditRequest when clicked', () => {
            let onEditRequestWasCalled = false;

            const props = {
                checklistItem: defaultChecklistItem,
                onEditRequest: () => {onEditRequestWasCalled = true;}
            };

            const button = enzymeHelper.findSingle(
                shallow(<ChecklistItem {...props} />),
                selector);

            button.simulate('click');
            expect(onEditRequestWasCalled).toBe(true);
        });
    });

    it('should render a status icon column', () => {
        const icon = enzymeHelper.findSingle(render(),
            'tbody > tr > td > ChecklistItemStatusIcon');

        expect(icon.props().status).toBe(defaultChecklistItem.status);
    });

    it('should render start/end time columns', () => {
        const clms = enzymeHelper.find(
            render(),
            'tbody > tr > td.timestamp > PlannedVsActual');

        expect(clms.length).toBe(2);

        let props = clms.at(0).props();
        expect(props.planned).toEqual(format.time(defaultChecklistItem.scheduledStartTime));
        expect(props.actual).toEqual(format.time(defaultChecklistItem.actualStartTime));

        props = clms.at(1).props();
        expect(props.planned).toEqual(format.time(defaultChecklistItem.scheduledEndTime));
        expect(props.actual).toEqual(format.time(defaultChecklistItem.actualEndTime));
    });

    it('should render a desription column', () => {
        const td = enzymeHelper.find(render(), 'tbody > tr > td').at(4);
        expect(td.text()).toEqual(defaultChecklistItem.description);
    });

    it('should render a "performed by" column', () => {
        const td = enzymeHelper.find(render(), 'tbody > tr > td').at(5);
        expect(td.text()).toEqual(defaultChecklistItem.userName);
    });
});
