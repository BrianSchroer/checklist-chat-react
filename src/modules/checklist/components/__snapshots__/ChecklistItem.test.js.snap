// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChecklistItem should render correctly 1`] = `
<tbody>
  <tr
    className="checklist-item-row"
  >
    <td>
      <button
        className="btn btn-primary btn-sm checklist-item-button"
        onClick={[Function]}
        title="Edit this item"
      >
        1
      </button>
    </td>
    <td>
      <i
        className="glyphicon glyphicon-unchecked"
        title="not started"
      />
    </td>
    <td
      className="timestamp"
    >
      <div
        className="planned-vs-actual"
      >
        <div
          className="planned-value-inaccurate"
        >
          8:00 am
        </div>
        <div
          className="actual-value"
        >
          8:03 am
        </div>
      </div>
    </td>
    <td
      className="timestamp"
    >
      <div
        className="planned-vs-actual"
      >
        <div
          className="planned-value-inaccurate"
        >
          9:00 am
        </div>
        <div
          className="actual-value"
        >
          9:30 am
        </div>
      </div>
    </td>
    <td>
      item description
    </td>
    <td>
      Jabba Script
    </td>
  </tr>
</tbody>
`;
